src/geoarrow.cpp

```diff
@@ -71,7 +71,178 @@ py::array_t<PyObjectGeography> from_geoarrow(py::object input,
+                        "are not supported");
+                }
+            }
+        }
+
+        ArrowArray* c_array = static_cast<ArrowArray*>(malloc(sizeof(ArrowArray)));
+        ArrowSchema* c_schema = static_cast<ArrowSchema*>(malloc(sizeof(ArrowSchema)));
+        move(&array_, &schema_, c_array, c_schema);
+
+        constexpr auto array_cleanup = [](void* ptr) noexcept {
+            auto array = static_cast<ArrowArray*>(ptr);
+            if (array->release != nullptr) {
+                array->release(array);
+            }
+            free(array);
+        };
+        py::capsule array_capsule{c_array, "arrow_array", array_cleanup};
+
+        constexpr auto schema_cleanup = [](void* ptr) noexcept {
+            auto schema = static_cast<ArrowSchema*>(ptr);
+            if (schema->release != nullptr) {
+                schema->release(schema);
+            }
+            free(schema);
+        };
+        py::capsule schema_capsule{c_schema, "arrow_schema", schema_cleanup};
+
+        return py::make_tuple(schema_capsule, array_capsule);
+    }
+
+    void move(ArrowArray* array_src,
+              ArrowSchema* schema_src,
+              ArrowArray* array_dst,
+              ArrowSchema* schema_dst) {
+        memcpy(array_dst, array_src, sizeof(struct ArrowArray));
+        array_src->release = nullptr;
+
+        memcpy(schema_dst, schema_src, sizeof(struct ArrowSchema));
+        schema_src->release = nullptr;
+    }
+
+    /// \brief Call data's release callback if valid
+    void reset() {
+        if (array_.release != nullptr) {
+            array_.release(&array_);
+        }
+
+        if (schema_.release != nullptr) {
+            schema_.release(&schema_);
+        }
+    }
+
+    /// \brief Call data's release callback if valid and move ownership of the data
+    /// pointed to by data
+    void reset(ArrowArray* array_src, ArrowSchema* schema_src) {
+        reset();
+        move(array_src, schema_src, &array_, &schema_);
+    }
+
+    ~ArrowArrayHolder() {
+        reset();
+    }
+
+protected:
+    ArrowArray array_;
+    ArrowSchema schema_;
+};
+
+ArrowArrayHolder to_geoarrow(py::array_t<PyObjectGeography> input,
+                             py::object output_schema,
+                             //  py::object geometry_encoding,
+                             bool planar,
+                             float tessellate_tolerance,
+                             int precision) {
+    ArrowArrayHolder array = ArrowArrayHolder();
+
+    s2geog::geoarrow::Writer writer;
+    std::vector<std::unique_ptr<s2geog::Geography>> s2geog_vec;
+
+    s2geog::geoarrow::ExportOptions options;
+    options.set_precision(precision);
+    if (planar) {
+        auto tol = S1Angle::Radians(tessellate_tolerance / EARTH_RADIUS_METERS);
+        options.set_tessellate_tolerance(tol);
+        // TODO make this configurable
+        // options.set_projection(s2geog::geoarrow::mercator());
```

Comment:
@benbovy we want to make this configurable, so the user can specify one of the pre-defined projections in s2geography (see https://github.com/paleolimbot/s2geography/pull/39 for the options in s2geograhy), but any thoughts on the user API to do this?

We could have a very shallow `Projection` class that essentially just holds a S2::Projection, and then some creation functions for the different projections. 

For example something with class methods like `spherely.Projection.pseudo_mercator()` ? Or just functions like `spherely.pseudo_mercator_projection()`? (or `spherely.projection_pseudo_mercator()`)

Reply:
(and R `s2` has `s2_projection_plate_carree()`, `s2_projection_mercator()`, `s2_projection_orthographic(centre=(0,0))` , listed on https://r-spatial.github.io/s2/reference/wk_handle.s2_geography.html)

Reply:
I like the shallow `Projection` class with class methods.

Reply:
I have a follow-up PR exposing projections using a shallow `spherely.Projection` class -> https://github.com/benbovy/spherely/pull/69